File: tests/fake-resend-server.ts
import { createServer, type Server } from 'node:http';
import type { AddressInfo } from 'node:net';
import type { ResendEmail, SendEmailInput } from '@/lib/email';

export class FakeResendServer {
  private server: Server | null = null;
  private sequence = 0;
  private readonly idempotentResponses = new Map<string, string>();
  private readonly idempotentBatchResponses = new Map<
    string,
    { body: string; ids: string[] }
  >();
  failNextSendStatus: number | null = null;
  failNextBatchStatus: number | null = null;
  failNextBatchCode = 'application_error';
  disconnectAfterNextBatch = false;
  malformedNextBatchResponse = false;
  sentMetadataFailuresRemaining = 0;
  sentMetadataRequestCount = 0;
  readonly sends: Array<{
    id: string;
    idempotencyKey: string | undefined;
    input: SendEmailInput;
  }> = [];
  readonly batches: Array<{
    idempotencyKey: string | undefined;
    inputs: SendEmailInput[];
    ids: string[];
  }> = [];
  readonly received = new Map<string, ResendEmail>();

  async start(url: string) {
    const target = new URL(url);
    this.server = createServer(async (request, response) => {
      try {
        await this.handle(request, response);
      } catch (error) {
        response.writeHead(500, { 'content-type': 'application/json' });
        response.end(JSON.stringify({ error: String(error) }));
      }
    });
    await new Promise<void>((resolve, reject) => {
      this.server?.once('error', reject);
      this.server?.listen(Number(target.port), target.hostname, () =>
        resolve(),
      );
    });
    const address = this.server.address() as AddressInfo;
    if (address.port !== Number(target.port)) {
      throw new Error('Fake Resend server started on an unexpected port');
    }
  }

  reset() {
    this.sequence = 0;
    this.sends.length = 0;
    this.batches.length = 0;
    this.idempotentResponses.clear();
    this.idempotentBatchResponses.clear();
    this.failNextSendStatus = null;
    this.failNextBatchStatus = null;
    this.failNextBatchCode = 'application_error';
    this.disconnectAfterNextBatch = false;
    this.malformedNextBatchResponse = false;
    this.sentMetadataFailuresRemaining = 0;
    this.sentMetadataRequestCount = 0;
    this.received.clear();
    this.received.set('em_received123', {
      id: 'em_received123',
      message_id: '<received123@example.com>',
      from: 'external@example.com',
      to: ['inbox@example.com'],
      subject: 'Received Email',
      created_at: '2026-07-19T03:52:03.099Z',
      text: 'Inbound test body',
      html: '<p>Inbound test body</p>',
      headers: { from: 'External Person <external@example.com>' },
      reply_to: [],
    });
  }

  async close() {
    if (!this.server) {
      return;
    }
    await new Promise<void>((resolve, reject) => {
      this.server?.close((error) => (error ? reject(error) : resolve()));
    });
    this.server = null;
  }

  private async handle(
    request: import('node:http').IncomingMessage,
    response: import('node:http').ServerResponse,
  ) {
    const url = new URL(request.url ?? '/', 'http://localhost');
    if (request.method === 'POST' && url.pathname === '/emails/batch') {
      const body = await readBody(request);
      const inputs = JSON.parse(body) as SendEmailInput[];
      if (this.failNextBatchStatus) {
        const status = this.failNextBatchStatus;
        const code = this.failNextBatchCode;
        this.failNextBatchStatus = null;
        this.failNextBatchCode = 'application_error';
        return json(response, status, {
          name: code,
          message: 'simulated_batch_failure',
        });
      }
      const idempotencyKey = request.headers['idempotency-key'] as
        | string
        | undefined;
      const existing = idempotencyKey
        ? this.idempotentBatchResponses.get(idempotencyKey)
        : undefined;
      if (existing && existing.body !== body) {
        return json(response, 409, {
          name: 'invalid_idempotent_request',
          message: 'batch payload changed',
        });
      }

      const ids = existing?.ids ?? inputs.map(() => `batch-${++this.sequence}`);
      if (!existing) {
        this.batches.push({ idempotencyKey, inputs, ids });
        for (const [index, input] of inputs.entries()) {
          this.sends.push({ id: ids[index], idempotencyKey, input });
        }
        if (idempotencyKey) {
          this.idempotentBatchResponses.set(idempotencyKey, { body, ids });
        }
      }
      if (this.disconnectAfterNextBatch) {
        this.disconnectAfterNextBatch = false;
        response.destroy();
        return;
      }
      if (this.malformedNextBatchResponse) {
        this.malformedNextBatchResponse = false;
        return json(response, 200, { data: ids.map(() => ({})) });
      }
      return json(response, 200, { data: ids.map((id) => ({ id })) });
    }

    if (request.method === 'POST' && url.pathname === '/emails') {
      const body = await readBody(request);
      const input = JSON.parse(body) as SendEmailInput;
      if (this.failNextSendStatus) {
        const status = this.failNextSendStatus;
        this.failNextSendStatus = null;
        return json(response, status, { error: 'simulated_send_failure' });
      }
      const idempotencyKey = request.headers['idempotency-key'] as
        | string
        | undefined;
      const existingId = idempotencyKey
        ? this.idempotentResponses.get(idempotencyKey)
        : undefined;
      const id = existingId ?? `sent-${++this.sequence}`;
      if (!existingId) {
        this.sends.push({ id, idempotencyKey, input });
        if (idempotencyKey) {
          this.idempotentResponses.set(idempotencyKey, id);
        }
      }
      return json(response, 200, { id });
    }

    const receivedMatch = url.pathname.match(/^\/emails\/receiving\/([^/]+)$/);
    if (request.method === 'GET' && receivedMatch) {
      const email = this.received.get(decodeURIComponent(receivedMatch[1]));
      return email
        ? json(response, 200, email)
        : json(response, 404, { error: 'not_found' });
    }

    const sentMatch = url.pathname.match(/^\/emails\/([^/]+)$/);
    if (request.method === 'GET' && sentMatch) {
      this.sentMetadataRequestCount++;
      if (this.sentMetadataFailuresRemaining > 0) {
        this.sentMetadataFailuresRemaining--;
        return json(response, 404, { error: 'not_ready' });
      }
      const id = decodeURIComponent(sentMatch[1]);
      const sent = this.sends.find((entry) => entry.id === id);
      return sent
        ? json(response, 200, {
            id,
            message_id: `<${id}@resend.test>`,
            from: sent.input.from,
            to: sent.input.to,
            subject: sent.input.subject,
            created_at: new Date().toISOString(),
            text: sent.input.text ?? null,
            html: sent.input.html ?? null,
          })
        : json(response, 404, { error: 'not_found' });
    }

    return json(response, 404, { error: 'not_found' });
  }
}

function json(
  response: import('node:http').ServerResponse,
  status: number,
  body: unknown,
) {
  response.writeHead(status, { 'content-type': 'application/json' });
  response.end(JSON.stringify(body));
}

async function readBody(
  request: import('node:http').IncomingMessage,
): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const chunk of request) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  }
  return Buffer.concat(chunks).toString('utf8');
}

File: src/test/kotlin/com/castab/resend/support/FakeResendServer.kt
package com.castab.resend.support

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import org.http4k.core.Method
import org.http4k.core.Response
import org.http4k.core.Status
import org.http4k.routing.bind
import org.http4k.routing.path
import org.http4k.routing.routes
import org.http4k.server.Http4kServer
import org.http4k.server.Jetty
import org.http4k.server.asServer
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/** Mirrors Resend's documented retrieve format, e.g. `2026-04-03 22:13:42.674981+00`. */
private val PROVIDER_TIMESTAMP = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSSSSS'+00'")

private fun providerTimestamp(): String =
    OffsetDateTime.now(ZoneOffset.UTC).format(PROVIDER_TIMESTAMP)

/** In-process stand-in for the Resend API, mirroring the Express `FakeResendServer`. */
class FakeResendServer {
    private val jsonCodec = Json { ignoreUnknownKeys = true }
    private var sequence = 0
    private val idempotentResponses = mutableMapOf<String, String>()
    private val idempotentBatchResponses = mutableMapOf<String, Pair<String, List<String>>>()

    data class Send(val id: String, val idempotencyKey: String?, val input: JsonObject)

    val sends = mutableListOf<Send>()
    val batches = mutableListOf<Triple<String?, List<JsonObject>, List<String>>>()
    val received = mutableMapOf<String, JsonObject>()

    var failNextSendStatus: Int? = null
    var failNextBatchStatus: Int? = null
    var failNextBatchCode: String = "application_error"
    var malformedNextBatchResponse: Boolean = false

    /** While true, every `GET /emails/{id}` retrieval fails with a 500 (simulates provider outage). */
    var failSentRetrievals: Boolean = false

    private lateinit var server: Http4kServer
    val baseUrl: String get() = "http://localhost:${server.port()}"

    fun start() {
        reset()
        server = handler().asServer(Jetty(0)).start()
    }

    fun stop() {
        if (::server.isInitialized) server.stop()
    }

    fun reset() {
        sequence = 0
        sends.clear()
        batches.clear()
        idempotentResponses.clear()
        idempotentBatchResponses.clear()
        failNextSendStatus = null
        failNextBatchStatus = null
        failNextBatchCode = "application_error"
        malformedNextBatchResponse = false
        failSentRetrievals = false
        received.clear()
        received["em_received123"] = buildJsonObject {
            put("id", JsonPrimitive("em_received123"))
            put("message_id", JsonPrimitive("<received123@example.com>"))
            put("from", JsonPrimitive("external@example.com"))
            put("to", JsonArray(listOf(JsonPrimitive("inbox@example.com"))))
            put("subject", JsonPrimitive("Received Email"))
            put("created_at", JsonPrimitive("2026-07-19T03:52:03.099Z"))
            put("text", JsonPrimitive("Inbound test body"))
            put("html", JsonPrimitive("<p>Inbound test body</p>"))
            put("headers", buildJsonObject { put("from", JsonPrimitive("External Person <external@example.com>")) })
            put("reply_to", JsonArray(emptyList()))
        }
    }

    /** Registers a retrievable received-email fixture (call after `reset`). */
    fun addReceived(id: String, messageId: String, from: String = "external@example.com", references: String? = null) {
        received[id] = buildJsonObject {
            put("id", JsonPrimitive(id))
            put("message_id", JsonPrimitive(messageId))
            put("from", JsonPrimitive(from))
            put("to", JsonArray(listOf(JsonPrimitive("inbox@example.com"))))
            put("subject", JsonPrimitive("Received Email"))
            put("created_at", JsonPrimitive("2026-07-19T03:52:03.099Z"))
            put("text", JsonPrimitive("Inbound test body"))
            put("html", JsonPrimitive("<p>Inbound test body</p>"))
            put("headers", buildJsonObject {
                put("from", JsonPrimitive(from))
                if (references != null) put("references", JsonPrimitive(references))
            })
            put("reply_to", JsonArray(emptyList()))
        }
    }

    private fun json(status: Int, body: JsonObject): Response =
        Response(Status(status, null)).header("content-type", "application/json").body(body.toString())

    private fun handler() = routes(
        "/emails/batch" bind Method.POST to { request ->
            val body = request.bodyString()
            val inputs = (jsonCodec.parseToJsonElement(body) as JsonArray).map { it as JsonObject }
            failNextBatchStatus?.let { status ->
                val code = failNextBatchCode
                failNextBatchStatus = null
                failNextBatchCode = "application_error"
                return@to json(status, buildJsonObject {
                    put("name", JsonPrimitive(code))
                    put("message", JsonPrimitive("simulated_batch_failure"))
                })
            }
            val idempotencyKey = request.header("idempotency-key")
            val existing = idempotencyKey?.let { idempotentBatchResponses[it] }
            if (existing != null && existing.first != body) {
                return@to json(409, buildJsonObject {
                    put("name", JsonPrimitive("invalid_idempotent_request"))
                    put("message", JsonPrimitive("batch payload changed"))
                })
            }
            val ids = existing?.second ?: inputs.map { "batch-${++sequence}" }
            if (existing == null) {
                batches.add(Triple(idempotencyKey, inputs, ids))
                inputs.forEachIndexed { index, input -> sends.add(Send(ids[index], idempotencyKey, input)) }
                if (idempotencyKey != null) idempotentBatchResponses[idempotencyKey] = body to ids
            }
            if (malformedNextBatchResponse) {
                malformedNextBatchResponse = false
                return@to json(200, buildJsonObject { put("data", buildJsonArray { ids.forEach { add(buildJsonObject {}) } }) })
            }
            json(200, buildJsonObject { put("data", buildJsonArray { ids.forEach { add(buildJsonObject { put("id", JsonPrimitive(it)) }) } }) })
        },
        "/emails" bind Method.POST to { request ->
            val body = request.bodyString()
            val input = jsonCodec.parseToJsonElement(body) as JsonObject
            failNextSendStatus?.let { status ->
                failNextSendStatus = null
                return@to json(status, buildJsonObject { put("error", JsonPrimitive("simulated_send_failure")) })
            }
            val idempotencyKey = request.header("idempotency-key")
            val existingId = idempotencyKey?.let { idempotentResponses[it] }
            val id = existingId ?: "sent-${++sequence}"
            if (existingId == null) {
                sends.add(Send(id, idempotencyKey, input))
                if (idempotencyKey != null) idempotentResponses[idempotencyKey] = id
            }
            json(200, buildJsonObject { put("id", JsonPrimitive(id)) })
        },
        "/emails/receiving/{id}" bind Method.GET to { request ->
            val id = request.path("id").orEmpty()
            received[id]?.let { json(200, it) } ?: json(404, buildJsonObject { put("error", JsonPrimitive("not_found")) })
        },
        "/emails/{id}" bind Method.GET to { request ->
            if (failSentRetrievals) {
                return@to json(500, buildJsonObject { put("error", JsonPrimitive("simulated_retrieval_failure")) })
            }
            val id = request.path("id").orEmpty()
            val sent = sends.firstOrNull { it.id == id }
            if (sent == null) {
                json(404, buildJsonObject { put("error", JsonPrimitive("not_found")) })
            } else {
                json(200, buildJsonObject {
                    put("id", JsonPrimitive(id))
                    put("message_id", JsonPrimitive("<$id@resend.test>"))
                    put("from", sent.input["from"] ?: JsonPrimitive(""))
                    put("to", sent.input["to"] ?: JsonArray(emptyList()))
                    put("subject", sent.input["subject"] ?: JsonPrimitive(""))
                    // Resend's retrieve response uses a Postgres-style timestamp, not an ISO instant.
                    put("created_at", JsonPrimitive(providerTimestamp()))
                    put("text", sent.input["text"] ?: kotlinx.serialization.json.JsonNull)
                    put("html", sent.input["html"] ?: kotlinx.serialization.json.JsonNull)
                })
            }
        },
    )
}
